File: backend/src/domain/assets.ts
import { DeleteCommand, GetCommand, PutCommand } from '@aws-sdk/lib-dynamodb';
import { docClient, queryAll, TABLE, toEntity } from '../lib/ddb.js';
import { conflict, notFound } from '../lib/errors.js';
import { newAssetCode, newId } from '../lib/ids.js';
import { assetCodeLookupPk, GSI2, PK, PREFIX, SK } from './keys.js';
import type { Asset, AssetStatus } from './types.js';

export interface AssetInput {
  title: string;
  category?: string;
  code?: string;
  creator?: string;
  identifier?: string;
  location?: string;
  condition?: string;
  replacementCost?: number;
  tags?: string[];
  status?: AssetStatus;
}

/** Label prefix hints the category at a glance on a printed shelf tag. */
const CODE_PREFIXES: Record<string, string> = {
  book: 'BK',
  equipment: 'EQ',
  tool: 'TL',
  media: 'MD',
  device: 'DV',
};

function toItem(asset: Asset) {
  return {
    PK: PK(asset.orgId),
    SK: SK.asset(asset.assetId),
    entityType: 'Asset',
    gsi2pk: assetCodeLookupPk(asset.orgId, asset.code),
    gsi2sk: `ASSET#${asset.assetId}`,
    ...asset,
  };
}

export async function listAssets(orgId: string): Promise<Asset[]> {
  const items = await queryAll<Asset>({
    TableName: TABLE(),
    KeyConditionExpression: 'PK = :pk AND begins_with(SK, :sk)',
    ExpressionAttributeValues: { ':pk': PK(orgId), ':sk': PREFIX.asset },
  });
  return items.sort((a, b) => a.title.localeCompare(b.title));
}

export async function getAsset(orgId: string, assetId: string): Promise<Asset> {
  const result = await docClient().send(
    new GetCommand({ TableName: TABLE(), Key: { PK: PK(orgId), SK: SK.asset(assetId) } }),
  );
  if (!result.Item) throw notFound('Asset not found');
  return toEntity<Asset>(result.Item);
}

export async function findAssetByCode(orgId: string, code: string): Promise<Asset | undefined> {
  const items = await queryAll<Asset>({
    TableName: TABLE(),
    IndexName: GSI2,
    KeyConditionExpression: 'gsi2pk = :pk',
    ExpressionAttributeValues: { ':pk': assetCodeLookupPk(orgId, code) },
  });
  return items[0];
}

/**
 * Scanners hand us whatever the QR encoded. We accept the deep-link URL we
 * print, a raw asset id, or a typed-in label code, so a single lookup path
 * serves the camera and the keyboard.
 */
export async function resolveAssetRef(orgId: string, ref: string): Promise<Asset> {
  const trimmed = ref.trim();
  if (!trimmed) throw notFound('No asset reference supplied');

  const fromUrl = trimmed.match(/\/a\/([^/?#]+)/);
  const candidate = fromUrl?.[1] ?? trimmed;

  if (candidate.includes('-')) {
    const byCode = await findAssetByCode(orgId, candidate);
    if (byCode) return byCode;
  }

  try {
    return await getAsset(orgId, candidate);
  } catch {
    const byCode = await findAssetByCode(orgId, candidate);
    if (byCode) return byCode;
    throw notFound(`No asset matches "${ref}" in this branch`);
  }
}

export async function createAsset(orgId: string, input: AssetInput): Promise<Asset> {
  const category = input.category?.trim() || 'book';
  const code = (input.code?.trim() || newAssetCode(CODE_PREFIXES[category] ?? 'AS')).toUpperCase();

  const clash = await findAssetByCode(orgId, code);
  if (clash) {
    throw conflict('DUPLICATE_CODE', `Label code ${code} is already used by "${clash.title}".`);
  }

  const now = new Date().toISOString();
  const asset: Asset = {
    orgId,
    assetId: newId(),
    code,
    title: input.title.trim(),
    category,
    creator: input.creator?.trim() || undefined,
    identifier: input.identifier?.trim() || undefined,
    location: input.location?.trim() || undefined,
    condition: input.condition ?? 'good',
    replacementCost: input.replacementCost,
    tags: input.tags,
    status: input.status ?? 'available',
    timesBorrowed: 0,
    createdAt: now,
    updatedAt: now,
  };

  await docClient().send(
    new PutCommand({
      TableName: TABLE(),
      Item: toItem(asset),
      ConditionExpression: 'attribute_not_exists(SK)',
    }),
  );
  return asset;
}

export async function updateAsset(
  orgId: string,
  assetId: string,
  patch: Partial<AssetInput>,
): Promise<Asset> {
  const current = await getAsset(orgId, assetId);

  if (patch.code && patch.code.toUpperCase() !== current.code) {
    const clash = await findAssetByCode(orgId, patch.code);
    if (clash && clash.assetId !== assetId) {
      throw conflict('DUPLICATE_CODE', `Label code ${patch.code} is already in use.`);
    }
  }

  if (patch.status && current.status === 'checked_out' && patch.status !== 'checked_out') {
    // Status is owned by the checkout transaction while an item is on loan.
    throw conflict(
      'ASSET_ON_LOAN',
      `"${current.title}" is on loan. Check it in (or mark the loan lost) to change its status.`,
    );
  }

  const next: Asset = {
    ...current,
    ...Object.fromEntries(Object.entries(patch).filter(([, v]) => v !== undefined)),
    code: patch.code ? patch.code.toUpperCase() : current.code,
    orgId,
    assetId,
    createdAt: current.createdAt,
    updatedAt: new Date().toISOString(),
  } as Asset;

  await docClient().send(new PutCommand({ TableName: TABLE(), Item: toItem(next) }));
  return next;
}

export async function deleteAsset(orgId: string, assetId: string): Promise<void> {
  const asset = await getAsset(orgId, assetId);
  if (asset.status === 'checked_out') {
    throw conflict('ASSET_ON_LOAN', `"${asset.title}" is on loan and cannot be deleted.`);
  }
  await docClient().send(
    new DeleteCommand({ TableName: TABLE(), Key: { PK: PK(orgId), SK: SK.asset(assetId) } }),
  );
}

File: backend/src/domain/assets.py
from datetime import datetime, timezone

from boto3.dynamodb.conditions import Key

from domain import keys
from domain.types import stock_levels, with_stock
from lib.ddb import drop_none, query_all, table, to_entity
from lib.errors import conflict, not_found
from lib.ids import new_asset_code, new_id

CODE_PREFIXES = {"book": "BK", "equipment": "EQ", "tool": "TL", "media": "MD", "device": "DV"}


def _now():
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.000Z")


def _item(asset: dict) -> dict:
    return drop_none(
        {
            "PK": keys.pk(asset["orgId"]),
            "SK": keys.sk_asset(asset["assetId"]),
            "entityType": "Asset",
            "gsi2pk": keys.asset_code_lookup_pk(asset["orgId"], asset["code"]),
            "gsi2sk": f"ASSET#{asset['assetId']}",
            **asset,
        }
    )


def list_assets(org_id: str) -> list:
    items = query_all(
        {
            "KeyConditionExpression": Key("PK").eq(keys.pk(org_id)) & Key("SK").begins_with(keys.PREFIX_ASSET),
        }
    )
    return sorted((with_stock(a) for a in items), key=lambda a: a.get("title", ""))


def get_asset(org_id: str, asset_id: str) -> dict:
    result = table().get_item(Key={"PK": keys.pk(org_id), "SK": keys.sk_asset(asset_id)})
    item = result.get("Item")
    if not item:
        raise not_found("Asset not found")
    return with_stock(to_entity(item))


def find_asset_by_code(org_id: str, code: str):
    items = query_all(
        {
            "IndexName": keys.GSI2,
            "KeyConditionExpression": Key("gsi2pk").eq(keys.asset_code_lookup_pk(org_id, code)),
        }
    )
    return with_stock(items[0]) if items else None


def resolve_asset_ref(org_id: str, ref: str) -> dict:
    trimmed = (ref or "").strip()
    if not trimmed:
        raise not_found("No asset reference supplied")
    import re

    from_url = re.search(r"/a/([^/?#]+)", trimmed)
    candidate = from_url.group(1) if from_url else trimmed
    if "-" in candidate:
        by_code = find_asset_by_code(org_id, candidate)
        if by_code:
            return by_code
    try:
        return get_asset(org_id, candidate)
    except Exception:
        by_code = find_asset_by_code(org_id, candidate)
        if by_code:
            return by_code
        raise not_found(f'No asset matches "{ref}" in this branch') from None


def persist_stock_fields(asset: dict) -> dict:
    """Write stock/available if an older unique-copy row is missing them."""
    if asset.get("stock") is not None and asset.get("available") is not None:
        return with_stock(asset)
    stock, available = stock_levels(asset)
    table().update_item(
        Key={"PK": keys.pk(asset["orgId"]), "SK": keys.sk_asset(asset["assetId"])},
        UpdateExpression="SET stock = if_not_exists(stock, :s), available = if_not_exists(available, :a)",
        ExpressionAttributeValues={":s": stock, ":a": available},
    )
    return with_stock({**asset, "stock": stock, "available": available})


def create_asset(org_id: str, input_data: dict) -> dict:
    category = (input_data.get("category") or "book").strip() or "book"
    code = ((input_data.get("code") or "").strip() or new_asset_code(CODE_PREFIXES.get(category, "AS"))).upper()
    clash = find_asset_by_code(org_id, code)
    if clash:
        raise conflict("DUPLICATE_CODE", f'Label code {code} is already used by "{clash["title"]}".')
    now = _now()
    stock = int(input_data["stock"]) if input_data.get("stock") is not None else 1
    asset = drop_none(
        {
            "orgId": org_id,
            "assetId": new_id(),
            "code": code,
            "sku": code,
            "title": input_data["title"].strip(),
            "category": category,
            "creator": (input_data.get("creator") or "").strip() or None,
            "identifier": (input_data.get("identifier") or "").strip() or None,
            "location": (input_data.get("location") or "").strip() or None,
            "condition": input_data.get("condition") or "good",
            "replacementCost": input_data.get("replacementCost"),
            "tags": input_data.get("tags"),
            "status": input_data.get("status") or "available",
            "stock": stock,
            "available": stock,
            "timesBorrowed": 0,
            "createdAt": now,
            "updatedAt": now,
        }
    )
    table().put_item(Item=_item(asset), ConditionExpression="attribute_not_exists(SK)")
    return with_stock(asset)


def update_asset(org_id: str, asset_id: str, patch: dict) -> dict:
    current = get_asset(org_id, asset_id)
    if patch.get("code") and patch["code"].upper() != current["code"]:
        clash = find_asset_by_code(org_id, patch["code"])
        if clash and clash["assetId"] != asset_id:
            raise conflict("DUPLICATE_CODE", f'Label code {patch["code"]} is already in use.')
    if patch.get("status") and current.get("unitsOnLoan", 0) > 0 and patch["status"] not in ("available", "checked_out"):
        raise conflict(
            "ASSET_ON_LOAN",
            f'"{current["title"]}" still has {current["unitsOnLoan"]} unit(s) on loan. Check them in first.',
        )
    current_stock, current_available = stock_levels(current)
    on_loan = current_stock - current_available
    next_stock = int(patch["stock"]) if patch.get("stock") is not None else current_stock
    if next_stock < on_loan:
        raise conflict(
            "STOCK_BELOW_LOANS",
            f"{on_loan} unit(s) are out on loan. Stock cannot be lower than that.",
        )
    next_asset = {**current, **{k: v for k, v in patch.items() if v is not None}}
    next_asset.update(
        {
            "code": patch["code"].upper() if patch.get("code") else current["code"],
            "sku": patch["code"].upper() if patch.get("code") else current.get("sku") or current["code"],
            "stock": next_stock,
            "available": next_stock - on_loan,
            "orgId": org_id,
            "assetId": asset_id,
            "createdAt": current["createdAt"],
            "updatedAt": _now(),
        }
    )
    next_asset.pop("unitsOnLoan", None)
    table().put_item(Item=_item(next_asset))
    return with_stock(next_asset)


def delete_asset(org_id: str, asset_id: str) -> None:
    asset = get_asset(org_id, asset_id)
    if asset.get("unitsOnLoan", 0) > 0:
        raise conflict(
            "ASSET_ON_LOAN",
            f'"{asset["title"]}" still has {asset["unitsOnLoan"]} unit(s) on loan and cannot be deleted.',
        )
    table().delete_item(Key={"PK": keys.pk(org_id), "SK": keys.sk_asset(asset_id)})
